### ISSUE TITLE
Add rule S6329: Allowing public network access to cloud resources is security-sensitive (APPSEC-169)

### ISSUE THREAD

**Issue by gaetan-ferry-sonarsource** [SonarSource/SonarJS#3460]
# WHAT
This rule update tries to detect AWS resources created with public Internet exposure.

## Instance-related constructs
For `aws-cdk-lib.aws_ec2.Instances`, sensitive if: 
- The constructor is called with the vpcSubnet attribute set and the subnet selection contains a public subnet.

Depending on the way the selection is built, we might not be able to know the restriction level of the subnet. We are sure when: 
- `vpcSubnet` is set with the `subnetType` property set to `ec2.SubnetType.PUBLIC`. 

If the appropriate properties for the `SubnetSelection` can not be resolved, do not raise an issue to avoid false positives. The `SubnetSelection` type accepts multiple properties. However, only `subnetType` allows for precise detection.

**Example**
```javascript
import {aws_ec2 as ec2} from 'aws-cdk-lib'
const instance = new ec2.Instance(this, "vpcSubnetPublic", {
    instanceType: nanoT2,
    machineImage: ec2.MachineImage.latestAmazonLinux(),
    vpc: vpc,
    vpcSubnets: {subnetType: ec2.SubnetType.PUBLIC} // Sensitive
})

const privateInstance = new ec2.Instance(
    this,
    "vpcSubnetPrivate", {
    instanceType: nanoT2,
    machineImage: ec2.MachineImage.latestAmazonLinux(),
    vpc: vpc,
    vpcSubnets: {subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS} // Compliant
})

const defaultPrivateInstance = new ec2.Instance( // Compliant, private by default
    this,
    "vpcSubnetPrivateDefault", {
    instanceType: nanoT2,
    machineImage: ec2.MachineImage.latestAmazonLinux(),
    vpc: vpc
})
```

For `aws-cdk-lib.aws_ec2.CfnInstance`:
**(1)** Raise an issue when:
- The constructor is called and the `networkInterfaces` property is a non-empty array that contains an `ec2.CfnInstance.NetworkInterfaceProperty` with associatePublicIpAddress property set to true

**(2)** Do not raise when the network interface specification places the interface in a private vpc subnet:

- the `subnetId` attribute of the `ec2.CfnInstance.NetworkInterfaceProperty` is set to the id of a private subnet. This can be detected when the id is obtained from a call to `aws-cdk-lib.aws_ec2.Vpc.selectSubnets()` with the `subnetType` parameter set to `aws-cdk-lib.aws_ec2.SubnetType.PRIVATE_ISOLATED, PRIVATE_WITH_EGRESS, or PRIVATE_WITH_NAT`.

When **(2)** can not be resolved and **(1)** is true, raise an issue anyway.

There are many ways to build a `subnetId` value. Only `selectSubnets` calls with `subnetType` set to PUBLIC allow for precise detection.

**Example**
```javascript
import {aws_ec2 as ec2} from 'aws-cdk-lib'

new ec2.CfnInstance(this, "cfnPublicExposed", {
    instanceType: "t2.micro",
    imageId: "ami-0ea0f26a6d50850c5",
    networkInterfaces: [
        {
            deviceIndex: "0",
            associatePublicIpAddress: true, // Sensitive
            deleteOnTermination: true,
            subnetId: vpc.selectSubnets({subnetType: ec2.SubnetType.PUBLIC}).subnetIds[0]
        }
    ]
})

new ec2.CfnInstance(this, "cfnPrivateWithPublicIp", {
    instanceType: "t2.micro",
    imageId: "ami-0ea0f26a6d50850c5",
    networkInterfaces: [
        {
            deviceIndex: "0",
            associatePublicIpAddress: true, // Compliant, the public IP won't be routable
            deleteOnTermination: true,
            subnetId: vpc.selectSubnets({subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS}).subnetIds[0]
        }
    ]
})

new ec2.CfnInstance(this, "cfnPrivate", {
    instanceType: "t2.micro",
    imageId: "ami-0ea0f26a6d50850c5",
    networkInterfaces: [
        {
            deviceIndex: "0",
            associatePublicIpAddress: false, // Compliant
            deleteOnTermination: true,
            subnetId: vpc.selectSubnets({subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS}).subnetIds[0]
        }
    ]
})
```

## Database instance related constructs
For `aws-cdk-lib.aws_rds.DatabaseInstance`:

Raise an issue when the constructor is called and:

- the `vpcSubnets` attribute is an instance of `aws-cdk-lib.aws_ec2.SubnetSelection` that selects a public subnet **(1)**
- AND the `publiclyAccessible` property is set to `true` or not set

**(1)** can be determined with precision only if the `subnetType` property of the `SubnetSelection` object is set to `aws-cdk-lib.aws_ec2.SubnetType.PUBLIC`

Also raise an issue when the constructor is called and:
- **(1)** Can not be determined precisely
- AND the `publiclyAccessible` property is set to `true`

![image{caption=Sorry for this image that uses Python style snake case}](https://user-images.githubusercontent.com/112399173/194551688-353a064d-4132-4af5-90e4-da8a05ccf142.png)

**Example**
```javascript
import {
    aws_ec2 as ec2,
    aws_rds as rds
} from 'aws-cdk-lib'

new rds.DatabaseInstance(this, "public_public_subnet", {
    engine: rds.DatabaseInstanceEngine.POSTGRES,
    vpc: vpc,
    publiclyAccessible: true, // Sensitive
    vpcSubnets: {subnetType: ec2.SubnetType.PUBLIC},
    securityGroups: [sg]
})

new rds.DatabaseInstance(this, "private_public_subnet", {
    engine: rds.DatabaseInstanceEngine.POSTGRES,
    vpc: vpc,
    publiclyAccessible: false, // Compliant
    vpcSubnets: {subnetType: ec2.SubnetType.PUBLIC},
    securityGroups: [sg]
})

new rds.DatabaseInstance(this, "public_private_subnet", {
    engine: rds.DatabaseInstanceEngine.POSTGRES,
    vpc: vpc,
    publiclyAccessible: true, // Compliant, IP won't be routable
    vpcSubnets: {subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS},
    securityGroups: [sg]
})
```

For `aws-cdk-lib.aws_rds.CfnDBInstance`:

Raise an issue when the constructor is called and: 
- the `publiclyAccessible` property is set to `true` **(1)**
- AND the db subnet is set to a public subnet **(2)**.

If **(1)** is true and **(2)** can not be determined, raise an issue anyway.

In the constructor call, **(2)** materializes only by the `dbSubnetGroupName` property which contains the name of a `CfnDBSubnetGroup`. The actual Public / Private information is only available in the `CfnDBInstance` constructor. If it is not available, **(2)** can not be determined. In the `CfnDBInstance` constructor, the Public / Private state can only be determined from the `subnetIds` property. If this one is the result of a call to `Vpc.selectSubnet` with the `subnetType` property set, then the Public / Private state can be determined:
- `aws-cdk-lib.aws_ec2.SubnetType.PUBLIC` => public
- `aws-cdk-lib.aws_ec2.SubnetType.PRIVATE_ISOLATED`,`PRIVATE_WITH_NAT`,`PRIVATE_WITH_EGRESS` => private

Otherwise, **(2)** can not be determined.
![image](https://user-images.githubusercontent.com/112399173/194553447-3399b35c-5e4a-450e-9199-b3c0395ad60a.png)

**Example**

```javascript
import {
    aws_ec2 as ec2,
    aws_rds as rds
} from 'aws-cdk-lib'

const rdsSubnetGroupPublic = new rds.CfnDBSubnetGroup(this, "publicSubnet", {
    dbSubnetGroupDescription: "Subnets",
    dbSubnetGroupName: "publicSn",
    subnetIds: vpc.selectSubnets({
        subnetType: ec2.SubnetType.PUBLIC
    }).subnetIds
})

const rdsSubnetGroupPrivate = new rds.CfnDBSubnetGroup(this, "privateSubnet",{
    dbSubnetGroupDescription: "Subnets",
    dbSubnetGroupName: "privateSn",
    subnetIds: vpc.selectSubnets({
        subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS
    }).subnetIds
})

new rds.CfnDBInstance(this, "public-public-subnet", {
    engine: "postgres",
    masterUsername: "foobar",
    masterUserPassword: "12345678",
    dbInstanceClass: "db.r5.large",
    allocatedStorage: "200",
    iops: 1000,
    dbSubnetGroupName: rdsSubnetGroupPublic.ref,
    publiclyAccessible: true, // Sensitive
    vpcSecurityGroups: [sg.securityGroupId]
})

new rds.CfnDBInstance(this, "private-private-subnet", {
    engine: "postgres",
    masterUsername: "foobar",
    masterUserPassword: "12345678",
    dbInstanceClass: "db.r5.large",
    allocatedStorage: "200",
    iops: 1000,
    dbSubnetGroupName: rdsSubnetGroupPrivate.ref,
    publiclyAccessible: false, // Compliant
    vpcSecurityGroups: [sg.securityGroupId]
})
```
## Database migration related constructs

For `aws-cdk-lib.aws_dms.CfnReplicationInstance`:

Raise an issue when:
- the `CfnReplicationInstance` constructor is called with the `publiclyAccessible` property is set to `true`
- the `CfnReplicationInstance` constructor is called and the `publiclyAccessible` property is not set

**Example**
```javascript
import {aws_dms as dms} from 'aws-cdk-lib'

new dms.CfnReplicationInstance(
    this, "explicitPublic", {
    replicationInstanceClass: "dms.t2.micro",
    allocatedStorage: 5,
    publiclyAccessible: true, // Sensitive
    replicationSubnetGroupIdentifier: subnetGroup.replicationSubnetGroupIdentifier,
    vpcSecurityGroupIds: [vpc.vpcDefaultSecurityGroup]
})

new dms.CfnReplicationInstance(
    this, "explicitPrivate", {
    replicationInstanceClass: "dms.t2.micro",
    allocatedStorage: 5,
    publiclyAccessible: false, // Compliant
    replicationSubnetGroupIdentifier: subnetGroup.replicationSubnetGroupIdentifier,
    vpcSecurityGroupIds: [vpc.vpcDefaultSecurityGroup]
})

new dms.CfnReplicationInstance( // Sensitive, public exposure is the default
    this, "defaultPublic", {
    replicationInstanceClass: "dms.t2.micro",
    allocatedStorage: 5,
    replicationSubnetGroupIdentifier: subnetGroup.replicationSubnetGroupIdentifier,
    vpcSecurityGroupIds: [vpc.vpcDefaultSecurityGroup]
})
```

# RSPEC
This [RSPEC PR](https://github.com/SonarSource/rspec/pull/1318) provides guidelines regarding messages and highlights in the Highlight and Message sections.

**Comment by pierre-loup-tristant-sonarsource** [comment]
Implementation LGTM
